File: Interval/3440. Reschedule Meetings for Maximum Free Time II/solution.py
from sortedcontainers import SortedList
class Solution:
    def maxFreeTime(self, eventTime: int, startTime: List[int], endTime: List[int]) -> int:
        startTime += [eventTime]
        endTime += [eventTime]
        n = len(startTime)

        gaps = SortedList()
        gaps_after = SortedList()
        for i in range(n-1):
            gaps_after.add(startTime[i+1] - endTime[i])

        res = prev = 0
        for i in range(n):
            duration = endTime[i] - startTime[i]
            gap = startTime[i] - prev
            
            if i+1<n:
                gaps_after.remove(startTime[i+1]-endTime[i])

            if gaps.bisect_left(duration) < len(gaps) or gaps_after.bisect_left(duration) < len(gaps_after):
                res = max(res, gap + duration + (startTime[i+1]-endTime[i] if i+1<n else 0))
            else:
                res = max(res, gap + (startTime[i+1]-endTime[i] if i+1<n else 0))

            gaps.add(gap)
            prev = endTime[i]
        return res
    
# optimization
class Solution:
    def maxFreeTime(self, eventTime: int, startTime: List[int], endTime: List[int]) -> int:
        gaps = [startTime[i] - (endTime[i-1] if i > 0 else 0) for i in range(len(startTime))]
        gaps.append(eventTime - endTime[-1])
        
        prefix_max = [0]
        for gap in gaps[:-1]:
            prefix_max.append(max(prefix_max[-1], gap))
        
        suffix_max = [0]
        for gap in reversed(gaps[1:]):
            suffix_max.append(max(suffix_max[-1], gap))
        suffix_max = suffix_max[::-1]
        
        max_gap = max(gaps)
        for i in range(len(startTime)):
            duration = endTime[i] - startTime[i]
            left_gap = gaps[i]
            right_gap = gaps[i+1]
            max_gap = max(max_gap, left_gap + right_gap)
            if max(prefix_max[i], suffix_max[i+1]) >= duration:
                max_gap = max(max_gap, left_gap + right_gap + duration)
        return max_gap

File: Interval/3440. Reschedule Meetings for Maximum Free Time II/solution.cpp
class Solution {
public:
    int maxFreeTime(int eventTime, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size();
        
        // Edge case: no events
        if (n == 0) return eventTime;
        
        // Step 1: Calculate gaps between consecutive events
        vector<int> gaps;
        gaps.reserve(n + 1); // Reserve space for efficiency
        
        // First gap: from 0 to first event start
        gaps.push_back(startTime[0]);
        
        // Middle gaps: between consecutive events
        for (int i = 1; i < n; i++) {
            gaps.push_back(startTime[i] - endTime[i-1]);
        }
        
        // Last gap: from last event end to total time
        gaps.push_back(eventTime - endTime[n-1]);
        
        // Step 2: Build prefix and suffix maximum arrays more efficiently
        vector<int> prefixMax(n + 2, 0);  // prefixMax[i] = max of gaps[0..i-1]
        for (int i = 0; i <= n; i++) {
            prefixMax[i + 1] = max(prefixMax[i], gaps[i]);
        }
        
        vector<int> suffixMax(n + 2, 0);  // suffixMax[i] = max of gaps[i..n]
        for (int i = n; i >= 0; i--) {
            suffixMax[i] = max(suffixMax[i + 1], gaps[i]);
        }
        
        // Step 3: Find maximum free time
        int res = *max_element(gaps.begin(), gaps.end());  // Strategy 1: largest single gap
        
        // Strategy 2: Try modifying each event
        for (int i = 0; i < n; i++) {
            int duration = endTime[i] - startTime[i];
            
            // Option 2a: Cancel event i (merge adjacent gaps)
            int mergedGap = gaps[i] + gaps[i + 1];
            res = max(res, mergedGap);
            
            // Option 2b: Move event i to another location
            // Find the largest gap outside of gaps[i] and gaps[i+1]            
            // Only move if the event can fit in another gap
            if (max(prefixMax[i], suffixMax[i + 2]) >= duration) {
                res = max(res, mergedGap + duration);
            }
        }
        
        return res;
    }
};

/* 
Visualization of the algorithm:

Original timeline:
[0] <gap1> [event1] <gap2> [event2] <gap3> [event3] <gap4> [eventTime]

Strategy 1: Use largest single gap
Answer = max(gap1, gap2, gap3, gap4)

Strategy 2a: Cancel an event (e.g., event2)
[0] <gap1> [event1] <gap2+gap3> [event3] <gap4> [eventTime]
Answer = gap2 + gap3

Strategy 2b: Move an event (e.g., move event2 to gap4 if it fits)
[0] <gap1> [event1] <gap2+gap3> [event3] <gap4-event2_duration> [event2] [eventTime]
Answer = gap2 + gap3 + event2_duration

The algorithm uses prefix/suffix max arrays to efficiently check if an event 
can be moved to any other gap without having to check each gap individually.
*/

/*
COMPLEXITY:
- Time: O(n) - single pass through events
- Space: O(n) - for gaps and prefix/suffix arrays

The algorithm correctly handles all three strategies:
1. Use largest existing gap
2. Cancel an event to merge adjacent gaps  
3. Move an event to create maximum continuous free time
*/